db.go

```diff
@@ -383,6 +383,22 @@ func (db *DB) compact() (changes bool, err error) {
 		if err := db.reload(); err != nil {
 			return changes, errors.Wrap(err, "reload blocks")
 		}
+		// After this is fixed https://github.com/prometheus/tsdb/issues/309,
+		// if the compaction of head did not create any new blocks (possible
+		// if the samples were deleted in that range), and if number of
+		// blocks are 0, db.reload() does not truncate head. Hence MinTime
+		// of head is not set properly. So only for that situation, we need
+		// to call db.head.Truncate(..) manually.
+		db.mtx.RLock()
+		l := len(db.blocks)
+		db.mtx.RUnlock()
+		if l == 0 {
+			err = db.head.Truncate(maxt)
+			if err != nil {
+				return changes, errors.Wrap(err, "head truncate failed (in compact)")
+
+			}
+		}
```

Comment:
why not move this logic inside:

```
func (db *DB) reload() (err error) {
....
if len(blocks) == 0 {
        // Truncate head and return
}
```

Reply:
To do this we need to pass the time at which head has to be truncated. I did that first by having something like a wrapper around `reload()` which takes the time. But making that change just for a special case in compaction looked like overkill.

Do you still want me to include it inside `reload()`?

Reply:
can't you use `mint, maxt := rangeForTimestamp(db.head.MinTime(), db.opts.BlockRanges[0])` within  `reload()` ?

Reply:
There is possibility of `reload()` being called in other places too. In that case if blocks were 0, we would end up compacting the head unnecessarily, don't you think so?

Reply:
yeah i still think we should avoid this edge case somehow. the size based retention PR https://github.com/prometheus/tsdb/pull/343 simplifies quite a few places so lets wait for that one to be merged and will revisit.